### PR TITLE
gh-114271: Make `_thread.ThreadHandle` thread-safe in free-threaded builds

### PR DESCRIPTION
**Pull request by mpage**
We protect the mutable state of `ThreadHandle` using a `_PyOnceFlag`. Concurrent operations (i.e. `join`) on `ThreadHandle` block until it is their turn to execute or an earlier operation succeeds. Once an operation has been applied successfully all future operations complete immediately.

The `join()` method is now idempotent. It may be called multiple times but the underlying OS thread will only be joined once. After `join()` succeeds, any future calls to `join()` will succeed immediately.

The `detach()` method has been removed; nothing was using it.

* Issue: gh-114271

### CODE REVIEW

**Comment by mpage** [comment]
@colesbury - Would you please have a look and add the "skip news" label to this? In order to limit the size of the changes I'd like to get this merged first and then stack the `tstate_lock` stuff on top.

File: Modules/_threadmodule.c
Comment (colesbury) [inline Modules/_threadmodule.c]: I don't think we want to call the `_PyOnceFlag` in the after fork handler. It's not necessary because only one thread survives a fork() and risks deadlock if some other thread was in the process of calling it.

File: Modules/_threadmodule.c
Comment (colesbury) [inline Modules/_threadmodule.c]: Let's move this check out to `ThreadHandle_join` before the `_PyOnceFlag_CallOnce`. Right now, if some other thread calls join() and then we call self.join() we'll block on the `_PyOnceFlag_CallOnce()` instead of immediately raising the error.

Comment (mpage) [reply Modules/_threadmodule.c]: Ah, good catch!

**Comment by mpage** [comment]
@colesbury - This should be good to go. Would you please have another look?

**Review by colesbury — COMMENTED** [review]
Oh, the thread id reuse is tricky.

I'd like to avoid adding `_PyEventRc` if possible. The event doesn't seem like a great fit for `ThreadHandleObject` because we are never waiting on the event -- we're just using it like an atomic flag.

What about something like:
1) Make `thread_is_exiting` a plain variable that we use atomic operations to access, like `int thread_is_exiting` (because we don't have atomic ops on `bool` currently)
2) Pass the `ThreadHandleObject` to `do_start_new_thread` instead of having another object with a separate lifetime.

**Comment by mpage** [comment]
> I'd like to avoid adding _PyEventRc if possible.

Me too! I thought pretty hard about this and couldn't come up with a better solution. Plus, I think we're going to need it later anyway if we want to be able to implement `join` with a timeout.

> What about something like: ...

I considered this, but I think we need to set the flag after the `PyThreadState` has been destroyed. Otherwise, you could end deadlocking in the following situation:

1. T0 joins T1.
2. T1 finishes. It has a thread-local with a finalizer that joins T1. Since we've set the flag before clearing the `PyThreadState`, we end up skipping the self-join check and deadlocking.

Since we have to set the flag after the `PyThreadState` has been destroyed, it's no long safe to call `Py_DECREF`.

**Review by colesbury — APPROVED** [review]
This looks good to me.

@pitrou, would you please look this over? The motivation is to make thread joining thread-safe in the freethreaded build, as well as fix a race condition that can affect both the default and freethreaded build (described by @mpage in #114839)

File: Include/internal/pycore_lock.h
Comment (pitrou) [inline Include/internal/pycore_lock.h]: Do these functions need to be inlined?

File: Modules/_threadmodule.c
Comment (pitrou) [inline Modules/_threadmodule.c:52]: Hmm, can we add at least a third-state for the non-joined non-detached state?

File: Modules/_threadmodule.c
Comment (pitrou) [inline Modules/_threadmodule.c]: It would be better IMHO for `state` to always hold a valid value (see comment about possible state values above).

File: Modules/_threadmodule.c
Comment (pitrou) [inline Modules/_threadmodule.c]: C99 has `_Bool`, but do we allow `bool` in the CPython source code? (I haven't followed, sorry)

File: Modules/_threadmodule.c
Comment (pitrou) [inline Modules/_threadmodule.c]: We should keep the `PyErr_SetString` here to make the error more informative.

File: Modules/_threadmodule.c
Comment (pitrou) [inline Modules/_threadmodule.c:82]: There are multiple members indicating progress/validity in this object: `state`, `is_valid`, `thread_is_exiting`.

IMHO it would be easier to reason about this with a single state member covering all (5?) possible states: invalid, running, exiting, joined, detached. It would also make sure we get transitions right, by adding runtime assertions.

(That said, the current logic looks ok)

File: Modules/_threadmodule.c
Comment (pitrou) [inline Modules/_threadmodule.c:1381]: Why not pass a `ThreadHandleObject*` directly? It's already ref-counted, and you could spare the additional ref-counted event since you're merely setting a flag...

File: Modules/_threadmodule.c
Comment (pitrou) [inline Modules/_threadmodule.c:221]: For the record, it seems this is only possible on Windows. On POSIX, the pthread handle will remain valid and unique until joined or detached.

Comment (pitrou) [reply Modules/_threadmodule.c:221]: Also, on Windows, it seems calling [GetExitCodeThread](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-getexitcodethread) allows knowing if a thread identified by its handle has exited.

So perhaps we can add this following abstraction function and remove the need for `thread_is_exiting`.
```c
bool PyThread_is_current(PyThread_handle_t th);

// POSIX
bool PyThread_is_current(PyThread_handle_t th) {
  return pthread_self() == (pthread_t) handle;
}

// Windows
bool PyThread_is_current(PyThread_handle_t th) {
  HANDLE hThread = (HANDLE) handle;
  if (GetThreadId(hThread) != GetCurrentThreadId()) {
    return false;
  }
  // The thread idents are equal, so either
  // 1. thread is the current thread
  // 2. thread was terminated and its ident reused for the current thread
  // Disambiguate by checking whether thread is still alive.
  unsigned exitcode;
  return GetExitCodeThread(hThread, &exitcode) && exitcode == STATUS_ACTIVE;
}
```

cc @gpshead

**Review by pitrou — COMMENTED** [review]
In general this looks ok but I think this can be simplified, see below.

File: Modules/_threadmodule.c
Comment (pitrou) [reply Modules/_threadmodule.c:1381]: Edit: ah, just saw the comment about finalization above. Darn!

File: Include/internal/pycore_lock.h
Comment (mpage) [reply Include/internal/pycore_lock.h]: I don't think so, but the implementation is quite short, so I don't think this hurts the overall readability of this file. That said, I don't feel strongly about it and am happy to move the implementation to `lock.c` if you want.

File: Modules/_threadmodule.c
Comment (mpage) [reply Modules/_threadmodule.c]: I believe the guidance is that it's fine to use as long as it won't leak into public headers.

File: Modules/_threadmodule.c
Comment (mpage) [reply Modules/_threadmodule.c]: Unless I'm misunderstanding I think that should already happen here. The call to `_PyOnceFlag` will call `detach_thread` if the handle has not already been joined/detached. `detach_thread` will call `PyErr_SetString` as before if the call to `PyThread_detach_thread` fails.

File: Modules/_threadmodule.c
Comment (pitrou) [reply Modules/_threadmodule.c]: Hmm, I see. Unfortunately it doesn't seem to be in PEP 7.

File: Modules/_threadmodule.c
Comment (mpage) [reply Modules/_threadmodule.c:221]: > For the record, it seems this is only possible on Windows. On POSIX, the pthread handle will remain valid and unique until joined or detached.

This affects POSIX systems too. The situation that we're preventing against here occurs once the handle has been joined or detached. Consider the following scenario with threads T0 and T1.

1. T0 starts, finishes, and the underlying pthread is detached.
2. T1 starts and is assigned the pthread handle used by T0. This is fine because the thread underlying T0 has been detached.
3. T1 attempts to join the handle for T0. Performing the self-join check here would fail because T1 has reused the handle for T0.

File: Include/internal/pycore_lock.h
Comment (pitrou) [reply Include/internal/pycore_lock.h]: I'm probably infected by C++, but putting implementations in a `.h` file tends to grow compile times and make headers less readable.

File: Modules/_threadmodule.c
Comment (pitrou) [reply Modules/_threadmodule.c]: Ah, right, sorry:

File: Modules/_threadmodule.c
Comment (pitrou) [reply Modules/_threadmodule.c:221]: > This affects POSIX systems too. The situation that we're preventing against here occurs once the handle has been joined or detached.

Ok, then can it simply be folded inside `join_thread`? Since you're calling `join_thread` with the once_flag, this won't be a problem anymore.

(but we may still need the is-current check for Windows)

File: Modules/_threadmodule.c
Comment (mpage) [reply Modules/_threadmodule.c:82]: We need `thread_is_exiting` but let me see if I can collapse `state` and `is_valid`. 

Another thing that would simplify this quite a bit is if we could remove the `detach` method. AFAICT it's only used in tests (to test the logic around attempting to join a detached handle). Since detaching only happens in the destructor, I think it should be safe to remove. What do you think?

Comment (pitrou) [reply Modules/_threadmodule.c:82]: Generally, we would rather not wait for the destructor to release resources, which is why `detach` is exposed (though it might not be called at higher levels currently?).

File: Modules/_threadmodule.c
Comment (colesbury) [reply Modules/_threadmodule.c]: I think it's covered by "Python 3.11 and newer versions use C11 without optional features." `bool` is available since C99 (and not optional).

File: Modules/_threadmodule.c
Comment (mpage) [reply Modules/_threadmodule.c:221]: > Ok, then can it simply be folded inside join_thread? 

No, that leaves us open to deadlocking in the following scenario:

1. T0 joins T1.
2. T1 self joins.

File: Modules/_threadmodule.c
Comment (mpage) [reply Modules/_threadmodule.c:82]: > though it might not be called at higher levels currently?

Yes, it doesn't appear to be called anywhere outside of tests. If nothing is using it I think we should delete it.

File: Modules/_threadmodule.c
Comment (pitrou) [reply Modules/_threadmodule.c]: `_Bool` is available as a standard C type, not `bool`: https://en.cppreference.com/w/c/language/type
But `bool` is a convenience macro defined in `stdbool.h`: https://en.cppreference.com/w/c/types

File: Modules/_threadmodule.c
Comment (pitrou) [reply Modules/_threadmodule.c:221]: Why wouldn't this prevented by the call_once flag? Edit: ah, ok, the once_flag *serializes* the calls to the inner function.

File: Modules/_threadmodule.c
Comment (pitrou) [reply Modules/_threadmodule.c:82]: If `Thread` had a `close` method for deterministic resource management (and/or supported the context manager protocol), it could call the `detach` method. But we can perhaps remove it for now, indeed.

**Comment by mpage** [comment]
@pitrou - I think I've addressed all of your comments except for the question about using `bool`. A quick grep of the codebase reveals quite a few uses of `bool` (in central places like `obmalloc.c` and `dictobject.c`, among others). Even if there isn't explicit guidance, there's certainly precedent. I've also removed the `detach()` method. Would you please have another look?

@colesbury - This has changed a bit since you reviewed it. Would you please have a look? In particular, I'd like a sanity check that the use of relaxed atomics for accessing handle state is OK.

**Comment by mpage** [comment]
The failing macos-13 builds appear unrelated to this PR. They're failing with errors like:

```
error: call to undeclared function 'is_pad'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
    if (py_is_pad(self->win)) {
```

File: Modules/_threadmodule.c
Comment (colesbury) [inline Modules/_threadmodule.c]: This can just be: `self->state = THREAD_HANDLE_INVALID`;

It's minor, but I generally prefer not using atomic operations where they're not necessary because it tends to hide data races from TSAN.

File: Modules/_threadmodule.c
Comment (colesbury) [inline Modules/_threadmodule.c]: This doesn't need synchronization because it's in the destructor. It should be okay to do something like:

```
if (self->state == THREAD_HANDLE_RUNNING) {
    if (PyThread_detach_thread(handle->handle)) {
        ...
    }
    ...
}
```

File: Modules/_threadmodule.c
Comment (colesbury) [inline Modules/_threadmodule.c]: I don't think the `THREAD_HANDLE_DETACHED` state is useful anymore. We only set it during destruction and then the handle is immediately freed.

**Review by colesbury — COMMENTED** [review]
A few minor comments below

File: Modules/_threadmodule.c
Comment (mpage) [reply Modules/_threadmodule.c]: Heh I had this and then changed it back. I figured it was better to use one pattern/abstraction everywhere, even when it isn't strictly required, to reduce cognitive load on future readers who might not be familiar with this code.

File: Modules/_threadmodule.c
Comment (mpage) [reply Modules/_threadmodule.c]: I have mixed feelings about this. I think there's some value in having the state reflect the transitions that the handle has gone through, even if it happens immediately before the object is freed, for debugging purposes. This is pretty hand-wavy, but in the past it has been super helpful to have extra data like this when groveling through core dumps.

Comment (colesbury) [reply Modules/_threadmodule.c]: Ok, it's fine to keep it then

File: Modules/_threadmodule.c
Comment (pitrou) [inline Modules/_threadmodule.c]: I don't know if this is logically ok, but a good rule of thumb is that the "relaxed" variants should only really be used for performance-critical code, which is not the case here.

File: Modules/_threadmodule.c
Comment (pitrou) [inline Modules/_threadmodule.c]: Can the comment elaborate a bit on the rationale for this?

File: Modules/_threadmodule.c
Comment (mpage) [reply Modules/_threadmodule.c]: I think we're OK here. The rationale being that we only need to access `state` atomically and don't thus don't need the ordering guarantees provided by the stronger memory orders. That said, I considered wrapping `ThreadHandle_join` in a critical section, but it feels a little heavy-handed.

I think it's also worth being painfully obvious that this is only relevant for the free-threaded build. There's no risk to correctness in the default build.

**Comment by mpage** [comment]
@colesbury - I think I've addressed your latest round of comments. Would you please have another look?

File: Modules/_threadmodule.c
Comment (pitrou) [reply Modules/_threadmodule.c]: I wouldn't really rely on "I think we're ok" given that we're talking about potentially hairy and difficult to reproduce issues (such as ordering of loads and stores on different CPUs).

Since this isn't performance-critical all, my vote vote is strongly in favor of the "normal" atomic API (i.e. sequentially consistent).

Comment (mpage) [reply Modules/_threadmodule.c]: > Since this isn't performance-critical all, my vote vote is strongly in favor of the "normal" atomic API (i.e. sequentially consistent).

Good point!

**Comment by mpage** [comment]
@pitrou - I think I've addressed everything now. Would you please have a look?

**Comment by mpage** [comment]
@pitrou @gpshead - Did you want to take a look at this?

File: Lib/test/test_thread.py
Comment (gpshead) [inline Lib/test/test_thread.py:192]: This is interesting... our docs already claim that threads can be joined multiple times.  So I wonder why this existing test logic was previously explicitly checking for an error here.

In this sense this change aligns with what our docs claim so a behavior change here could be seen as a bugfix.  I do not _expect_ anyone to be depending on subsequent join()s of a thread raising regardless.

File: Lib/test/test_thread.py
Comment (gpshead) [inline Lib/test/test_thread.py:279]: Replace `self_joiner_handle.join()` with something that records itself having been called.

```python
def self_joiner_recording_join():
    rv = orig_self_joiner_handler_orig_join()
    self_joiner_handler.join_recorded.set()
    return rv

self_joiner_handle_orig_join = self_joiner_handle.join
self_joiner_handle.join_recorded = threading.Event()
self_joiner_handle.join = self_joiner_recording_join
# ... do stuff, start the thread that is supposed to join it, etc.
self.assertTrue(
    self.joiner_handle.join_recorded.wait(2),
    msg="self_joiner thread not joined after 2 seconds.")
```

File: Lib/test/test_thread.py
Comment (gpshead) [inline Lib/test/test_thread.py:237]: from a readers point of view, it isn't clear why we ever would('ve) deadlock in this situation.  this smells like a regression test, it is useful for future maintainers to link to an issue or somewhere with information on when and how this happened.  otherwise the test could _look_ like an overly complicated way to test that a self join fails in need of simplification.

File: Lib/threading.py
Comment (gpshead) [inline Lib/threading.py:999]: Remove `self._join_lock = None` up in the constructor as this attribute is no longer used.

**Review by gpshead — CHANGES_REQUESTED** [review]
_(review in progress, more to come)_

File: Modules/_threadmodule.c
Comment (gpshead) [inline Modules/_threadmodule.c]: Explicitly assign values to these (non-0 to help differentiate from "uninitialized"?) so that people debugging crash dumps know for sure what is what by reading the code without having to understand what values the compiler auto-assigned to names. (I'm assuming C even specifies how auto-enumerating works -- 0 start and +1 each time IIRC -- but don't want anyone but a compiler maintainer to _need_ to know that)

File: Lib/test/test_thread.py
Comment (gpshead) [reply Lib/test/test_thread.py:237]: Referring readers to the comment within threadmodule.c ThreadHandle_join() for example.

File: Modules/_threadmodule.c
Comment (gpshead) [inline Modules/_threadmodule.c:139]: comments in this function as to why it is safe to not use the get/set_ functions for atomic access would be nice.

In general i believe this should never be possible to call if another thread had a reference and were calling another method that accessed them?

File: Modules/_threadmodule.c
Comment (gpshead) [inline Modules/_threadmodule.c:1518]: Do this before the Py_DECREF in case that triggers a destructors with horrible `__del__` python code doing things like poking at threading privates.  (defensive coding; cue _"nobody should ever do that and we do/can/should prevent that horror"_)

File: Modules/_threadmodule.c
Comment (gpshead) [inline Modules/_threadmodule.c:1275]: can you add a comment as to what the existing internal `struct bootstate` is used for?  it looks like how we "bootstrap" a new thread by passing it the parameters it needs to run?

File: Lib/test/test_thread.py
Comment (mpage) [reply Lib/test/test_thread.py:279]: Unfortunately this is still racy, though probably less error prone than the previous approach. To be correct, we need to wait until the joining thread is blocked, not just called, before unblocking the self-joining thread. I don't think it's possible to do that (without adding a lot of extra complexity into the runtime).

Comment (gpshead) [reply Lib/test/test_thread.py:279]: Ugh for sure.

There might be some creative^Wgross ways to figure that out via ctypes or testcapi? I dislike tests that sleep and hope for things to happen as it is best to assume they always wind up flaky and incur ongoing tweaking and random frustration.

Is it possible to introspect the thread at the OS level on one of our tier1 platforms to deterministically understand when it is actually blocked on the thing we want it blocked on?  (we'd be back to polling waiting for that to happen, but if it _is_ deterministic that would still be nice)

File: Modules/_threadmodule.c
Comment (mpage) [reply Modules/_threadmodule.c:1275]: > it looks like how we "bootstrap" a new thread by passing it the parameters it needs to run?

Yep, exactly.

File: Modules/_threadmodule.c
Comment (mpage) [reply Modules/_threadmodule.c:139]: > In general i believe this should never be possible to call if another thread had a reference

Yeah, this is the destructor so nothing else should have a reference and the refcount going to zero is a "syncronizes-with" event, so any change from another thread will be visible here.

File: Lib/test/test_thread.py
Comment (mpage) [reply Lib/test/test_thread.py:279]: > I dislike tests that sleep

Me too!

> Is it possible to introspect the thread at the OS level on one of our tier1 platforms to deterministically understand when it is actually blocked on the thing we want it blocked on?

This feels like it might end up being brittle in a different way. We'd end up relying on implementation details of `PyEvent` (i.e. putting the thread to sleep while blocked) and the OS.

Comment (mpage) [reply Lib/test/test_thread.py:279]: So unfortunately we can't even override `join` as it's read-only. I can do some digging into how difficult it would be to verify that the thread is actually blocked if you feel strongly, but I'm not sure it's worth the added complexity.

**Comment by mpage** [comment]
I have made the requested changes; please review again

**Comment by mpage** [comment]
@gpshead - Would you please have another look at this?

File: Lib/test/test_thread.py
Comment (gpshead) [reply Lib/test/test_thread.py:279]: makes sense, I like to _try_ to avoid timing related issues but sometimes how to do so is not clear cut.  we can revisit this test if it causes problems.

File: Modules/_threadmodule.c
Comment (gpshead) [inline Modules/_threadmodule.c:181]: I do hope nothing is depending on this, it's _supposed_ to be a very internal implementation detail.  we'll find out.  :)

**Review by gpshead — APPROVED** [review]

**Pull request merged**
